pkg/sql/colexec/execgen/template.go

```diff
@@ -401,6 +522,135 @@ func findTemplateFuncs(f *dst.File) map[string]*funcInfo {
+	if len(args) == 1 {
+		// Base case: transform the final options list into an arguments list of
+		// lists where each arguments list is a single element containing one of
+		// the final options.
+		// For example, given [[true, false]], we'll return:
+		// [[true], [false]]
+		ret := make([][]string, len(args[0]))
+		for i, arg := range args[0] {
+			ret[i] = []string{arg}
+		}
+		return ret
+	}
+	rest := generateInstantiateCombinations(args[1:])
+	ret := make([][]string, 0, len(rest)*len(args[0]))
+	for _, argOption := range args[0] {
+		// For every option of argument, prepend it to every args list from
+		// the recursive step.
+		for _, args := range rest {
+			ret = append(ret, append([]string{argOption}, args...))
+		}
+	}
+	return ret
+}
+
+// generateSwitchStatementLookup ...
+// remainingArgs is a list of lists of actual instantiations. For example, if
+// we had:
+// execgen:instantiate<red, potato>
+// execgen:instantiate<red, orange>
+// execgen:instantiate<yellow, orange>
+//
+// remainingArgs would be {{red, potato}, {red, orange}, {yellow orange}}
+func generateSwitchStatementLookup(
+	info *funcInfo, curArgs []dst.Expr, remainingTemplateParams []string, remainingArgs [][]string,
+) *dst.SwitchStmt {
+	ret := &dst.SwitchStmt{
+		Tag:  dst.NewIdent(remainingTemplateParams[0]),
+		Body: &dst.BlockStmt{},
+	}
+	defaultCase := &dst.CaseClause{
+		Body: []dst.Stmt{
+			mustParseStmt(`panic(fmt.Sprint("unknown value", ` + remainingTemplateParams[0] + `))`),
+		},
+	}
+	if len(remainingArgs[0]) == 1 {
+		// Base case. We finished switching on all template params, time to actually
+		// invoke the fully specialized function.
+		stmtList := make([]dst.Stmt, len(remainingArgs)+1)
+		for i := range remainingArgs {
+			argList := append(curArgs, dst.NewIdent(remainingArgs[i][0]))
+			call := &dst.CallExpr{
+				Fun:  dst.NewIdent(info.decl.Name.Name),
+				Args: argList,
+			}
+			_, call, _ = replaceTemplateVars(info, call)
+			var stmt dst.Stmt
+			if info.decl.Type.Results != nil {
+				stmt = &dst.ReturnStmt{Results: []dst.Expr{call}}
+			} else {
+				stmt = &dst.ExprStmt{X: call}
+			}
+			stmtList[i] = &dst.CaseClause{
+				List: []dst.Expr{dst.NewIdent(remainingArgs[i][0])},
+				Body: []dst.Stmt{stmt},
+			}
+		}
+		stmtList[len(stmtList)-1] = defaultCase
+		ret.Body.List = stmtList
+		return ret
+	}
+
+	// Recursive case: we have more args to deal with
+	groupedArgs := make(map[string][][]string)
+	for _, argList := range remainingArgs {
+		firstArg := argList[0]
+		groupedArgs[firstArg] = append(groupedArgs[firstArg], argList[1:])
+	}
+	stmtList := make([]dst.Stmt, len(groupedArgs)+1)
+	var i int
+	// Sort firstArgs lexicographically, so we have a consistent output order.
+	firstArgs := make([]string, 0, len(groupedArgs))
+	for firstArg := range groupedArgs {
+		firstArgs = append(firstArgs, firstArg)
+	}
+	sort.Strings(firstArgs)
+
+	for _, firstArg := range firstArgs {
```

Comment:
[nit] seems like you can just make this `for i, firstArg := range...` and remove the `i` declaration above, right?

Reply:
Good point, thanks